spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/Spark3Util.java

```diff
@@ -996,4 +1027,134 @@ public String unknown(
+          for (FileStatus fs : fileStatuses) {
+            map.put(fs.getPath(), fs);
+          }
+          cachedLeafFiles = map;
+        } catch (IOException e) {
+          throw new RuntimeException("error listing files for path=" + rootPath, e);
+        }
+      }
+      return cachedLeafFiles;
+    }
+
+    static List<FileStatus> listLeafDirs(
+        SparkSession spark, Path path, StructType partitionSpec, int level) throws IOException {
+      List<FileStatus> leafDirs = new ArrayList<>();
+      int numPartitionCols = partitionSpec.fields().length;
+      if (level < numPartitionCols) {
+        try (FileSystem fs = path.getFileSystem(spark.sparkContext().hadoopConfiguration())) {
+          List<FileStatus> dirs =
+              Stream.of(fs.listStatus(path))
+                  .filter(FileStatus::isDirectory)
+                  .collect(Collectors.toList());
+          for (FileStatus dir : dirs) {
+            // stop recursive call once we reach the expected end of partitions as per table schema
+            if (level == numPartitionCols - 1) {
+              leafDirs.add(dir);
+            } else {
+              leafDirs.addAll(listLeafDirs(spark, dir.getPath(), partitionSpec, level + 1));
+            }
+          }
+        }
+      }
+      return leafDirs;
+    }
+
+    @Override
+    public scala.collection.immutable.Map<Path, FileStatus[]> leafDirToChildrenFiles() {
+      if (cachedLeafDirToChildFiles == null) {
+        List<Tuple2<Path, FileStatus[]>> tuple2s =
+            JavaConverters.seqAsJavaList(leafFiles().values().toSeq()).stream()
+                .map(
+                    fileStatus -> {
+                      // Create an empty data file in the leaf dir.
+                      // As this index is only used to list partition directories,
+                      // we can stop listing the leaf dir to avoid unnecessary listing which can
+                      // take a while on folders with 1000s of files
+                      return new Tuple2<>(
+                          fileStatus.getPath(),
+                          new FileStatus[] {createEmptyChildDataFileStatus(fileStatus)});
+                    })
+                .collect(Collectors.toList());
+        cachedLeafDirToChildFiles =
+            (scala.collection.immutable.Map<Path, FileStatus[]>)
+                Map$.MODULE$.apply(JavaConverters.asScalaBuffer(tuple2s));
+      }
+      return cachedLeafDirToChildFiles;
+    }
+
+    private FileStatus createEmptyChildDataFileStatus(FileStatus fileStatus) {
+      return new FileStatus(
+          1L,
+          false,
+          fileStatus.getReplication(),
+          1L,
+          fileStatus.getModificationTime(),
+          fileStatus.getAccessTime(),
+          fileStatus.getPermission(),
+          fileStatus.getOwner(),
+          fileStatus.getGroup(),
+          new Path(fileStatus.getPath(), fileStatus.getPath().toString() + "/dummyDataFile"));
+    }
```

Comment:
maybe add a comment beside boolean parameter to help with reading? https://github.com/apache/iceberg/blob/master/CONTRIBUTING.md#boolean-arguments
```java
      return new FileStatus(
          1L,
          false, /* is directory */
          fileStatus.getReplication(),
          1L,
          fileStatus.getModificationTime(),
          fileStatus.getAccessTime(),
          fileStatus.getPermission(),
          fileStatus.getOwner(),
          fileStatus.getGroup(),
          new Path(fileStatus.getPath(), fileStatus.getPath().toString() + "/dummyDataFile"));
    }
```

Reply:
<img width="307" alt="image" src="https://user-images.githubusercontent.com/69539469/232584235-b3354106-a1ce-4a49-835b-3aeef349f232.png">